lib/elixir/lib/stream.ex

```diff
@@ -1383,6 +1383,16 @@ defmodule Stream do
         fn file -> File.close(file) end
       )
 
+      iex> Stream.resource(
+      ...>   fn -> 1 end,
+      ...>   fn
+      ...>     acc when acc > 5 -> {:halt, nil}
+      ...>     acc -> {[acc], acc+1}
+      ...>   end,
+      ...>   fn _ -> nil end
+      ...> ) |> Enum.to_list()
+      [1,2,3,4,5]
```

Comment:
The issue with this example is that it is unfortunately not a good for fit resources, as there is no resource here.  :/ Using unfold for this, for example, would work better. Thoughts?

Reply:
I wrote that after reading this code: https://github.com/danschultzer/pow/blob/master/guides/redis_cache_store_backend.md

It took me a while to understand the .resource function (and correct that code), as a functional concept. I was trying to explain it a little bit better than the current example. And there is no limitation of using it to resources only - and I think it's quite hard to think about a case that will at the same time:
* Will be using resource
* will be simple enough to show the functional essence of `resource`
* could be copy-pasted into iex without any extra steps

So I decided to make an elementary example :)

Reply:
Yes, I can see the pros of this example, but the cons is that it is misleading since there is no resource, so it can lead people to use resource when they should not. I am a bit torn but we probably shouldn't merge it as is, unfortunately. Thanks!

Reply:
Sure, I understand!

What do you think about example, where it will be provided as an explicit Resource-mock module?

Reply:
What do you mean?

Reply:
Could something like `StringIO` be a good example? It requires opening and closing and can also be looped through. It also is relatively simple compared to other things I thought about 😅

Reply:
I was thinking about something like:
```
   defmodule RemoteReader do
     defstruct [:data]
 
     require Logger
     def initialize(), do: %__MODULE__{data: [1,2,3,4,5]}
    
     def read_next(%__MODULE__{data: []} = reader), do: {:halt, reader}
     def read_next(%__MODULE__{data: [hd | tail]}), do: {[hd], %__MODULE__{data: tail}}
    
     def cleanup(%__MODULE__{data: data}), do: Logger.info("Remote read completed successfully, data left was #{inspect(data)}. Cleanup completed.")
   end
   
   Stream.resource(
     &RemoteReader.initialize/0,
     &RemoteReader.read_next/1,
     &RemoteReader.cleanup/1
   ) |> Enum.to_list()
  [1, 2, 3, 4, 5]
```

Reply:
Yeah, such module introduces too many concepts. Maybe the StringIO that @michalmuskala suggested?

Reply:
Sure, let me know what do you think:
```
Stream.resource(
  fn -> 
    {:ok, pid} = StringIO.open("string")
    pid
  end,
  fn pid -> 
    case IO.getn(pid, "", 1) do
      :eof -> {:halt, pid}
      char -> {[char], pid}
    end
  end,
  fn pid -> StringIO.close(pid) end
) |> Enum.to_list()
["s", "t", "r", "i", "n", "g"]
```

Reply:
I like that. In this case we could simply use a string, but I guess it is a good example of it being backed by a resource. Can you please update the PR accordingly?

Reply:
Sure, done 👍